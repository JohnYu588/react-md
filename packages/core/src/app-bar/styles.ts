import { cnb } from "cnbuilder";
import { type BoxOptions, box } from "../box/styles.js";
import { type BackgroundColor, cssUtils } from "../cssUtils.js";
import { type CssPosition } from "../types.js";
import { bem } from "../utils/bem.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-app-bar-height"?: string | number;
    "--rmd-app-bar-background-color"?: string;
    "--rmd-app-bar-color"?: string;
    "--rmd-app-bar-surface-background-color"?: string;
    "--rmd-app-bar-surface-color"?: string;
  }
}

const styles = bem("rmd-app-bar");
const titleStyles = bem("rmd-app-bar-title");

export type AppBarPosition = "top" | "bottom";

/**
 * - `"clear"` - the background color will be transparent
 * - `"primary"` - defaults to the current primary theme color
 * - `"secondary"` - defaults to the current secondary theme color
 * - `"surface"` - this will use the current surface background color which
 *   defaults to `#fff` for light themes, `#242424` for dark themes, and
 *   `#424242` for dark themes when the `$disable-dark-elevation` is set to
 *   `true`
 *
 * The default dark theme surface color also depends on the `$fixed-elevation`
 * value.
 */
export type AppBarTheme = BackgroundColor | "clear";

/**
 * - `"auto"` - the height will be determined by the height of the content
 * - `"normal"` - defaults to `3.5rem` (`$height`)
 * - `"prominent"` - defaults to `7rem` (`$prominent-height`)
 * - `"dense"` - defaults to `3rem` (`$dense-height`)
 * - `"prominent-dense"` - defaults to `6rem` (`$prominent-dense-height`)
 */
export type AppBarHeight =
  | "auto"
  | "normal"
  | "prominent"
  | "dense"
  | "prominent-dense";

/** @since 6.0.0 */
export interface AppBarClassNameOptions extends Omit<BoxOptions, "fullWidth"> {
  className?: string;

  /**
   * Set this to `"fixed"` or `"sticky"` to set `position: fixed;` or
   * `position: sticky;` to the app bar. The default position will be static and
   * inline with other content.
   *
   * @defaultValue `'static'`
   */
  position?: CssPosition;

  /**
   * The position within the page to "fix" the `AppBar` when the `fixed` prop is
   * enabled.
   *
   * @defaultValue `"top"`
   */
  pagePosition?: AppBarPosition;

  /**
   * Set this to `true` to remove the box-shadow.
   *
   * @defaultValue `false`
   */
  disableElevation?: boolean;

  /**
   * The theme to apply to the `AppBar`.
   *
   * @defaultValue `"primary"`
   * @see {@link AppBarTheme}
   */
  theme?: AppBarTheme;

  /**
   * @defaultValue `stacked ? "auto" : "normal"`
   * @see {@link AppBarHeight}
   */
  height?: AppBarHeight;

  /** @see {@link CssUtilsOptions.surfaceColor} */
  surfaceColor?: "light" | "dark";

  /**
   * Set this to `true` if the app bar's positioning and width should be
   * changed whenever the scrollbar is visible on the page. This defaults to
   * `true` when the {@link position} prop is `true` so that once dialogs and menus
   * become visible the contents in the app bar do not need to be repainted.
   *
   * @since 6.0.0
   * @defaultValue `position === "fixed"`
   */
  scrollbarOffset?: boolean;
}

/**
 * Apply the `className`s for a tree component. This will be type-safe if using
 * typescript.
 *
 * @since 6.0.0
 */
export function appBar(options: AppBarClassNameOptions = {}): string {
  const {
    className,
    theme = "primary",
    stacked,
    height = stacked ? "auto" : "normal",
    position = "static",
    pagePosition = "top",
    scrollbarOffset = position === "fixed",
    align,
    grid,
    gridColumns,
    gridName,
    justify,
    reversed,
    surfaceColor,
    disableWrap = true,
    disablePadding,
    disableElevation,
  } = options;
  const surface = theme === "surface";
  const clear = theme === "clear";

  return cnb(
    styles({
      surface,
      [height]: height !== "normal",
      fixed: position !== "static",
      stacked,
      sticky: position === "sticky",
      [pagePosition]: position !== "static",
      elevated: position !== "static" && !disableElevation,
      "scrollbar-offset": scrollbarOffset,
      "static-scrollbar-offset": position === "static" && scrollbarOffset,
    }),
    box({
      align,
      stacked,
      fullWidth: true,
      grid,
      gridColumns,
      gridName,
      justify,
      reversed,
      disableWrap,
      disablePadding,
    }),
    cssUtils({
      backgroundColor: !surface && !clear ? theme : undefined,
      surfaceColor,
    }),
    className
  );
}

/**
 * - `"small"` - the first character in the title will be `1rem` (`$keyline`)
 *   from the edge of the app bar horizontally
 * - `"nav"` - this should be set when there is a nav button before the title so
 *   that the first character in the title will be `4.5rem` (`title-keyline`)
 * - `"list"` - this should be used when the title should align with the list
 *   item keyline and there is no nav icon before.
 *
 * @since 6.0.0
 */
export type AppBarTitleKeyline = "small" | "nav" | "list";

/** @since 6.0.0 */
export interface AppBarTitleClassNameOptions {
  className?: string;

  /**
   * @defaultValue `"small"`
   * @see {@link AppBarTitleKeyline}
   */
  keyline?: AppBarTitleKeyline;
}

/**
 * Apply the `className`s for a `AppBarTitle` component. This will be type-safe
 * if using typescript.
 *
 * @since 6.0.0
 */
export function appBarTitle(options: AppBarTitleClassNameOptions = {}): string {
  const { className, keyline = "small" } = options;
  return cnb(
    titleStyles({
      keyline: keyline == "list",
      "nav-keyline": keyline === "nav",
    }),
    className
  );
}
