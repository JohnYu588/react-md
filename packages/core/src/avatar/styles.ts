import { cnb } from "cnbuilder";
import { bem } from "../utils/bem";

const styles = bem("rmd-avatar");

/** @remarks \@since 6.0.0 */
export interface AvatarClassNameOptions {
  className?: string;

  /** @defaultValue `""` */
  color?: string;

  /** @defaultValue `"avatar"` */
  size?: "avatar" | "icon";
}

/**
 * @remarks \@since 6.0.0
 */
export function avatar(options: AvatarClassNameOptions): string {
  const { className, color = "", size } = options;

  return cnb(styles({ [color]: color, icon: size === "icon" }), className);
}

/** @remarks \@since 6.0.0 */
export interface AvatarImageClassNameOptions {
  className?: string;
}

/**
 * @remarks \@since 6.0.0
 */
export function avatarImage(options: AvatarImageClassNameOptions = {}): string {
  const { className } = options;

  return cnb(styles("image"), className);
}
