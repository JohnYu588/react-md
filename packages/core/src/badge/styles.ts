import { cnb } from "cnbuilder";
import { type BackgroundColor, cssUtils } from "../cssUtils.js";
import { bem } from "../utils/bem.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-badge-size"?: string | number;
    "--rmd-badge-offset"?: string | number;
    "--rmd-badge-offset-top"?: string | number;
    "--rmd-badge-offset-right"?: string | number;
  }
}

const styles = bem("rmd-badge");

/**
 * @since 6.0.0 Renamed `"default"` to `"greyscale"` and added all
 * theme colors.
 */
export type BadgeTheme = BackgroundColor | "greyscale" | "clear";

/**
 * @since 6.0.0
 */
export interface BadgeClassNameOptions {
  className?: string;

  /** @defaultValue `""greyscale` */
  theme?: BadgeTheme;
}

/**
 * @since 6.0.0
 */
export function badge(options: BadgeClassNameOptions = {}): string {
  const { className, theme = "greyscale" } = options;
  const greyscale = theme === "greyscale";
  const clear = theme === "clear";

  return cnb(
    styles({ greyscale }),
    cssUtils({ backgroundColor: !clear && !greyscale ? theme : undefined }),
    className
  );
}
