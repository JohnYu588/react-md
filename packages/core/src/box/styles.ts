import { cnb } from "cnbuilder";
import { bem } from "../utils/bem.js";

const styles = bem("rmd-box");

declare module "react" {
  interface CSSProperties {
    "--rmd-box-gap"?: string | number;
    "--rmd-box-padding"?: string | number;
    "--rmd-box-item-min-size"?: string | number;
    "--rmd-box-columns"?: string | number;
  }
}

/**
 * @since 6.0.0
 */
export type BoxAlignItems =
  | "start"
  | "flex-start"
  | "center"
  | "end"
  | "flex-end"
  | "stretch";

/**
 * @since 6.0.0
 */
export type BoxJustifyContent =
  | BoxAlignItems
  | "space-around"
  | "space-between"
  | "space-evenly";

/**
 * @since 6.0.0
 */
export type BoxFlexDirection = "row" | "column";

/**
 * @since 6.0.0
 */
export interface BoxOptions {
  className?: string;

  /**
   * Set this to `true` to use `display: grid` instead of `display: flex`.
   *
   * @defaultValue `false`
   */
  grid?: boolean;

  /**
   * Set this to `true` to apply `width: 100%`. This can be useful when using
   * nested box layouts.
   *
   * @defaultValue `false`
   */
  fullWidth?: boolean;

  /**
   * Set this to `true` to set `flex-wrap: nowrap`.
   *
   * @defaultValue `false`
   */
  disableWrap?: boolean;

  /**
   * Set this to `true` to disable the default padding.
   *
   * @defaultValue `false`
   */
  disablePadding?: boolean;

  /**
   * This is a grid configuration
   * This should match one of the names in the `$grids` map. So for example:
   *
   * ```scss
   * @use "react-md" with (
   *   $grids: (
   *     small: (
   *       min: 5rem
   *     ),
   *     medium: (
   *       min: 7rem,
   *       gap: 0.5rem,
   *       padding: 2rem,
   *     ),
   *   ),
   * );
   * ```
   *
   * The `gridName` should be `"small" | "medium"`.
   *
   * @defaultValue `""`
   */
  gridName?: string;

  /**
   * @defaultValue `"fit"`
   */
  gridColumns?: "fit" | "fill" | number;

  /**
   * @defaultValue `""`
   */
  align?: BoxAlignItems;

  /**
   * The default value is really `center` or whatever the `$default-align-items` is set to.
   *
   * @defaultValue `""`
   */
  justify?: BoxJustifyContent;

  /**
   * Set this to `true` to set `flex-direction: column` which will stack all
   * items in the box.
   *
   * @defaultValue `false`
   */
  stacked?: boolean;

  /**
   * Set this to `true` to reverse the `flex-direction`. i.e.
   * - `flex-direction: row-reverse`
   * - `flex-direction: column-reverse`
   *
   * @defaultValue `false`
   */
  reversed?: boolean;
}

/**
 * @since 6.0.0
 */
export function box(options: BoxOptions = {}): string {
  const {
    className,
    align = "",
    grid = false,
    gridName = "",
    gridColumns = "fit",
    justify = "",
    stacked = false,
    reversed,
    fullWidth = false,
    disableWrap = false,
    disablePadding = false,
  } = options;

  return cnb(
    styles({
      wrap: !disableWrap,
      padded: !disablePadding,
      column: stacked && !reversed,
      reverse: !stacked && reversed,
      "column-reverse": stacked && reversed,
      "full-width": fullWidth,
      grid,
      "grid-fill": gridColumns === "fill",
      "grid-columns": typeof gridColumns === "number",
      [gridName]: grid && gridName,
      "align-start": align === "start" || align === "flex-start",
      "align-center": align === "center",
      "align-end": align === "end" || align === "flex-end",
      "align-stretch": align === "stretch",
      "justify-center": justify === "center",
      "justify-start": justify === "start" || justify === "flex-start",
      "justify-end": justify === "end" || justify === "flex-end",
      "justify-stretch": justify === "stretch",
      "justify-around": justify === "space-around",
      "justify-between": justify === "space-between",
      "justify-evenly": justify === "space-evenly",
    }),
    className
  );
}
