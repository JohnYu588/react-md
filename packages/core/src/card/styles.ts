import { cnb } from "cnbuilder";

import { cssUtils } from "../cssUtils.js";
import { bem } from "../utils/bem.js";

const cardStyles = bem("rmd-card");

/** @since 6.0.0 */
export interface CardClassNameOptions {
  className?: string;

  /**
   * Set this to `true` if the card should use apply an inset `border` instead
   * of `box-shadow`.
   *
   * @defaultValue `false`
   */
  bordered?: boolean;

  /**
   * Set this to `true` if the card should gain additional box shadow while
   * hovered.
   *
   * @defaultValue `false`
   */
  raisable?: boolean;

  /**
   * Set this to `true` if the card should gain the pointer cursor while
   * hovering and other interaction styles and not using the `ClickableCard`
   * component.
   *
   * @defaultValue `false`
   */
  interactable?: boolean;
}

/**
 * @since 6.0.0
 */
export function card(options: CardClassNameOptions = {}): string {
  const { className, bordered, raisable, interactable } = options;

  return cnb(
    cardStyles({
      bordered,
      shadowed: !bordered,
      raisable: !bordered && raisable,
      interactable,
    }),
    className
  );
}

/** @since 6.0.0 */
export interface CardHeaderClassNameOptions {
  className?: string;

  /** @defaultValue `false` */
  addonAfter?: boolean;

  /** @defaultValue `false` */
  addonBefore?: boolean;
}

/**
 * @since 6.0.0
 */
export function cardHeader(options: CardHeaderClassNameOptions = {}): string {
  const { className, addonAfter = false, addonBefore = false } = options;

  return cnb(
    cardStyles("header", {
      "addon-after": addonAfter && !addonBefore,
      "addon-before": addonBefore && !addonAfter,
      surrounded: addonAfter && addonBefore,
    }),
    className
  );
}

/** @since 6.0.0 */
export interface CardHeaderContentClassNameOptions {
  className?: string;
}

/**
 * @since 6.0.0
 */
export function cardHeaderContent(
  options: CardHeaderContentClassNameOptions = {}
): string {
  const { className } = options;

  return cnb(cardStyles("header-content"), className);
}

/** @since 6.0.0 */
export interface CardTitleClassNameOptions {
  className?: string;
}

/**
 * @since 6.0.0
 */
export function cardTitle(options: CardTitleClassNameOptions = {}): string {
  const { className } = options;

  return cnb(cardStyles("title"), className);
}

/** @since 6.0.0 */
export interface CardSubtitleClassNameOptions {
  className?: string;
}

/**
 * @since 6.0.0
 */
export function cardSubtitle(
  options: CardSubtitleClassNameOptions = {}
): string {
  const { className } = options;

  return cnb(cardStyles("subtitle"), className);
}

/** @since 6.0.0 */
export interface CardContentClassNameOptions {
  className?: string;

  /**
   * Set this to `true` to disable the card's padding.
   *
   * @defaultValue `false`
   */
  disablePadding?: boolean;

  /**
   * Set this to `true` to disable applying the `text-secondary-color`.
   *
   * @defaultValue `false`
   */
  disableSecondaryColor?: boolean;

  /**
   * Set this to `true` to disable applying extra `padding-bottom` if the
   * `CardContent` component is the last child.
   *
   * @defaultValue `false`
   * @since 6.0.0 Renamed from `disableExtraPadding`
   */
  disableLastChildPadding?: boolean;
}

/**
 * @since 6.0.0
 */
export function cardContent(options: CardContentClassNameOptions = {}): string {
  const {
    className,
    disablePadding,
    disableSecondaryColor,
    disableLastChildPadding,
  } = options;

  return cnb(
    cardStyles("content", {
      padded: !disablePadding,
      "padding-bottom": !disableLastChildPadding,
    }),
    cssUtils({
      textColor: !disableSecondaryColor ? "text-secondary" : undefined,
    }),
    className
  );
}

/** @since 6.0.0 */
export interface CardFooterClassNameOptions {
  className?: string;
}

/**
 * @since 6.0.0
 */
export function cardFooter(options: CardFooterClassNameOptions = {}): string {
  const { className } = options;

  return cnb(cardStyles("footer"), className);
}
