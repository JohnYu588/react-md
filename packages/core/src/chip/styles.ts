import { cnb } from "cnbuilder";
import { cssUtils } from "../cssUtils.js";
import { bem } from "../utils/bem.js";

const styles = bem("rmd-chip");

/**
 * @remarks \@since 6.0.0
 *
 * Note for the future: Once Firefox supports the `:has` selector, the
 * `leftAddon` and `rightAddon` can be removed. The styles would automatically
 * apply the padding instead:
 *
 * ```scss
 * &:has(.rmd-icon:first-child, .rmd-avatar:first-child) {
 *   @include core.auto-rtl(
 *     padding-left,
 *     $addon-left-padding,
 *     $horizontal-padding
 *   );
 * }
 *
 * $index: if(
 *   not core.$interaction-mode or core.$interaction-mode == press,
 *   1,
 *   2 // last child is always the `.rmd-ripple-container`
 * );
 * &:has(.rmd-icon:nth-last-child(#{$index})) {
 *   @include core.auto-rtl(
 *     padding-right,
 *     $addon-right-padding,
 *     $horizontal-padding
 *   );
 * }
 * ```
 */
export interface ChipClassNameOptions {
  className?: string;
  selectedClassName?: string;

  /** @defaultValue `"solid"` */
  theme?: "outline" | "solid";

  /** @defaultValue `false` */
  disabled?: boolean;

  /** @defaultValue `false` */
  selected?: boolean;

  /** @defaultValue `false` */
  selectedThemed?: boolean;

  /** @defaultValue `false` */
  noninteractive?: boolean;

  /** @defaultValue `false` */
  leftAddon?: boolean;

  /** @defaultValue `false` */
  rightAddon?: boolean;

  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * @remarks \@since 6.0.0
 */
export function chip(options: ChipClassNameOptions = {}): string {
  const {
    className,
    theme = "solid",
    pressed = false,
    disabled = false,
    selected = false,
    selectedThemed = false,
    selectedClassName,
    noninteractive = false,
    pressedClassName,
    leftAddon = false,
    rightAddon = false,
  } = options;

  return cnb(
    styles({
      themed: !disabled && selected && selectedThemed,
      selected: !disabled && selected && !selectedThemed,
      solid: theme === "solid",
      outline: theme === "outline",
      disabled,
      noninteractive,
      "left-addon": leftAddon && !rightAddon,
      "right-addon": !leftAddon && rightAddon,
      surrounded: leftAddon && rightAddon,
      pressed,
    }),
    selected && selectedClassName,
    pressedClassName,
    cssUtils({ textColor: disabled ? "text-disabled" : undefined }),
    className
  );
}

/**
 * @remarks \@since 6.0.0
 */
export interface ChipContentClassNameOptions {
  className?: string;
}

/**
 * @remarks \@since 6.0.0
 */
export function chipContent(options: ChipContentClassNameOptions = {}): string {
  const { className } = options;

  return cnb(styles("content"), className);
}
