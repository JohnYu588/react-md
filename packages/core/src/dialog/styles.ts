import { cnb } from "cnbuilder";

import {
  type CSSTransitionClassNames,
  type TransitionTimeout,
} from "../transition/types.js";
import { bem } from "../utils/bem.js";
import { DISPLAY_NONE_CLASS } from "../utils/isElementVisible.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-dialog-min-width"?: string | number;
    "--rmd-dialog-horizontal-margin"?: string | number;
    "--rmd-dialog-vertical-margin"?: string | number;
    "--rmd-dialog-z-index"?: string | number;
    "--rmd-dialog-header-padding"?: string | number;
    "--rmd-dialog-header-padding-bottom"?: string | number;
    "--rmd-dialog-content-padding"?: string | number;
    "--rmd-dialog-footer-padding"?: string | number;
    "--rmd-dialog-width"?: string | number;
    "--rmd-dialog-small-width"?: string | number;
    "--rmd-dialog-medium-width"?: string | number;
    "--rmd-dialog-large-width"?: string | number;
    "--rmd-dialog-extra-large-width"?: string | number;
  }
}

const styles = bem("rmd-dialog");
const containerStyles = bem("rmd-dialog-container");

export type DialogType = "full-page" | "centered" | "custom";

/**
 * This can be used to enforce a specific width for dialogs instead of relying
 * on the size of the content to determine the width. The width will also ensure
 * that it does not overflow based on the viewport width and margins applied.
 *
 * For example: if the `width="extra-large"` and the total viewport size is
 * `600px`, the dialog width would be `420px` since there is a default `80px`
 * margin to the left and right of the dialog. If the user expands the browser,
 * the dialog width will continue to grow until it reaches the `extra-large`
 * width and stop growing from that point.
 *
 * @since 6.0.0
 */
export type DialogWidth = "auto" | "small" | "medium" | "large" | "extra-large";

/**
 * @since 6.0.0
 */
export interface DialogContainerClassNameOptions {
  className?: string;

  centered?: boolean;
  displayNone?: boolean;
}

/**
 * @since 6.0.0
 */
export function dialogContainer(
  options: DialogContainerClassNameOptions = {}
): string {
  const { className, centered, displayNone } = options;

  return cnb(
    containerStyles({ centered }),
    displayNone && DISPLAY_NONE_CLASS,
    className
  );
}

/** @since 6.0.0 */
export interface DialogClassNameOptions {
  className?: string;

  /**
   * @defaultValue `"centered"`
   */
  type?: DialogType;

  /**
   * @defaultValue `"auto"`
   */
  width?: DialogWidth;

  /**
   * @defaultValue `false`
   */
  fixed?: boolean;

  /**
   * @defaultValue `type === "full-page"`
   */
  outline?: boolean;

  /**
   * This is mostly used for handling nested dialogs and removes any box shadow
   * on a dialog that has a child visible.
   *
   * @defaultValue `false`
   */
  disableBoxShadow?: boolean;
}

/** @since 6.0.0 */
export function dialog(options: DialogClassNameOptions = {}): string {
  const {
    type = "centered",
    width,
    fixed = false,
    outline = type === "full-page",
    disableBoxShadow,
    className,
  } = options;

  return cnb(
    styles({
      fixed,
      outline,
      centered: type === "centered",
      "full-page": type === "full-page",
      "no-box-shadow": type === "centered" && disableBoxShadow,
      "s-width": width === "small",
      "m-width": width === "medium",
      "l-width": width === "large",
      "xl-width": width === "extra-large",
    }),
    className
  );
}

/** @since 6.0.0 */
export interface DialogHeaderClassNameOptions {
  className?: string;
}

/**
 * NOTE: The default `DialogHeader` component uses the `Box` component so the
 * `box` class name utility function or `Box` component will probably be
 * required if using this util
 *
 * @since 6.0.0
 */
export function dialogHeader(
  options: DialogHeaderClassNameOptions = {}
): string {
  const { className } = options;

  return cnb(styles("header"), className);
}

/** @since 6.0.0 */
export interface DialogContentClassNameOptions {
  className?: string;

  /** @defaultValue `false` */
  disablePadding?: boolean;
}

/** @since 6.0.0 */
export function dialogContent(
  options: DialogContentClassNameOptions = {}
): string {
  const { className, disablePadding = false } = options;

  return cnb(styles("content", { padded: !disablePadding }), className);
}

/**
 * An optional alignment for the content within the footer. Since the majority
 * of dialog footers are used to contain action buttons, the default alignment
 * is near the end.
 *
 * @since 3.1.0
 */
export type DialogFooterAlignment =
  | "none"
  | "start"
  | "end"
  | "between"
  | "stacked-start"
  | "stacked-end";

/** @since 6.0.0 */
export interface DialogFooterClassNameOptions {
  className?: string;

  /** @defaultValue `"end"` */
  align?: DialogFooterAlignment;
}

/** @since 6.0.0 */
export function dialogFooter(
  options: DialogFooterClassNameOptions = {}
): string {
  const { align = "end", className } = options;

  return cnb(
    styles("footer", {
      flex: align !== "none",
      "flex-v": align === "stacked-start" || align === "stacked-end",
      start: align === "start" || align === "stacked-start",
      between: align === "between",
      end: align === "end" || align === "stacked-end",
    }),
    className
  );
}

/** @since 4.0.0 */
export const DEFAULT_DIALOG_TIMEOUT = {
  enter: 200,
  exit: 150,
} as const satisfies TransitionTimeout;

/** @since 4.0.0 */
export const DEFAULT_DIALOG_CLASSNAMES = {
  appear: "rmd-dialog--enter",
  appearActive: "rmd-dialog--enter-active",
  enter: "rmd-dialog--enter",
  enterActive: "rmd-dialog--enter-active",
  exit: "rmd-dialog--exit",
  exitActive: "rmd-dialog--exit-active",
} as const satisfies CSSTransitionClassNames;
