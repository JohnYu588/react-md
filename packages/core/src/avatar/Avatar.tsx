import { forwardRef, type HTMLAttributes, type ImgHTMLAttributes } from "react";
import { type PropsWithRef } from "../types.js";
import { avatar, avatarImage } from "./styles.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-avatar-background-color"?: string;
    "--rmd-avatar-color"?: string;
    "--rmd-avatar-border-color"?: string;
    "--rmd-avatar-border-radius"?: string | number;
    "--rmd-avatar-size"?: string | number;
    "--rmd-avatar-font-size"?: string | number;
  }
}

export type AvatarImgAttributes = ImgHTMLAttributes<HTMLImageElement>;

export interface AvatarProps extends HTMLAttributes<HTMLSpanElement> {
  /**
   * This should be an image `src` attribute to create an avatar from. When this
   * prop is defined, you should not add any children to the avatar as the
   * positioning will break.
   */
  src?: string;

  /**
   * An optional alt tag to display on the `<img>` when the `src` prop is also
   * applied.
   *
   * For accessibility and screen readers, you normally do not want to actually
   * provide this prop. This should only be used if the `Avatar` is not
   * accompanied by some other component or main content as it will be extra
   * noise for screen readers.
   *
   * @defaultValue `""`
   */
  alt?: string;

  /**
   * An optional `referrerPolicy` to provide to the `<img>` element if the `src`
   * or `imgProps` props are provided.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-referrerpolicy|Referrer Policy}
   *
   * @remarks \@since 2.2.0
   */
  referrerPolicy?: AvatarImgAttributes["referrerPolicy"];

  /**
   * An optional object of image props and ref that can be used to create an
   * image within the `Avatar`. This can be useful to add a custom `style`
   * or`className` to the `<img>` element if that additional customization is
   * needed.
   *
   * Note: The values in this object will override the `src`, `alt`, and
   * `referrerPolicy` root level avatar props if they exist on this object.
   *
   * @remarks \@since 2.2.0
   */
  imgProps?: PropsWithRef<AvatarImgAttributes, HTMLImageElement>;

  /**
   * An optional color to apply to the avatar. This will apply a className of
   * `rmd-avatar--${color}`, so only the keys from the `$rmd-avatar-colors` Map
   * are supported by default. It is recommended to create custom colors using
   * the `rmd-avatar-theme-update-var` mixin with custom class names if the
   * default colors aren't extensive enough.
   *
   * @defaultValue `""`
   */
  color?: string;

  /**
   * @defaultValue `"avatar"`
   */
  size?: "avatar" | "icon";
}

/**
 * **Server Component**
 *
 * An `Avatar` is generally used to represent objects or people within your app.
 * The avatar can consist of an image, an icon, or some text to display. When
 * the avatar is not an image, different themes can be applied to make the
 * avatar more unique.
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Avatar } from "@react-md/core";
 * import { type ReactElement } from "react";
 *
 * export function Example(): ReactElement {
 *   return <Avatar src="/path-to-img.png" />;
 * }
 * ```
 */
export const Avatar = forwardRef<HTMLSpanElement, AvatarProps>(
  function Avatar(props, ref) {
    const {
      className,
      children,
      src,
      size = "avatar",
      alt = "",
      color = "",
      imgProps,
      referrerPolicy,
      ...remaining
    } = props;

    let img;
    if (src || imgProps) {
      img = (
        <img
          src={src}
          alt={alt}
          referrerPolicy={referrerPolicy}
          {...imgProps}
          className={avatarImage({ className: imgProps?.className })}
        />
      );
    }

    return (
      <span
        {...remaining}
        ref={ref}
        role="presentation"
        className={avatar({ color, size, className })}
      >
        {img}
        {children}
      </span>
    );
  }
);
