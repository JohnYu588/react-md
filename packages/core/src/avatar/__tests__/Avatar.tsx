import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { render, screen } from "../../test-utils/index.js";

import { Avatar } from "../Avatar.js";
import { avatarImage } from "../styles.js";

describe("Avatar", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLSpanElement>();
    const props = {
      "data-testid": "avatar",
      ref,
    } as const;

    const { rerender } = render(<Avatar {...props} />);
    const avatar = screen.getByTestId("avatar");
    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
    expect(ref.current).toBe(avatar);
    expect(avatar).toMatchSnapshot();

    rerender(
      <Avatar
        {...props}
        style={{ color: "red" }}
        className="custom-class-name"
      />
    );
    expect(avatar).toMatchSnapshot();

    rerender(<Avatar {...props} color="primary" />);
    expect(avatar).toMatchSnapshot();
  });

  it("should render an img if a src or imgProps are provided", () => {
    const props = {
      "data-testid": "avatar",
      src: "/some-image.png",
    } as const;
    const { rerender } = render(<Avatar {...props} />);

    const avatar = screen.getByTestId("avatar");
    const img = screen.getByRole("img");
    expect(img).toHaveAttribute("src", props.src);
    expect(img).toHaveAttribute("alt", "");
    expect(avatar).toMatchSnapshot();

    rerender(<Avatar {...props} referrerPolicy="no-referrer" />);
    expect(img).toHaveAttribute("referrerPolicy", "no-referrer");
    expect(avatar).toMatchSnapshot();

    const ref = createRef<HTMLImageElement>();
    rerender(
      <Avatar
        {...props}
        imgProps={{
          ref,
          src: "/another-image.png",
          style: { opacity: 0.5 },
          className: "custom-class-name",
        }}
      />
    );
    expect(ref.current).toBeInstanceOf(HTMLImageElement);
    expect(ref.current).toBe(img);
    expect(img).toHaveAttribute("src", "/another-image.png");
    expect(avatar).toMatchSnapshot();
  });

  describe("styling utils", () => {
    it("should allow the avatarImage to be called without any parameters", () => {
      expect(avatarImage()).toMatchSnapshot();
    });
  });
});
