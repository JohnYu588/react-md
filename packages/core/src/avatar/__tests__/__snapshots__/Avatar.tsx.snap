// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<span
  class="rmd-avatar"
  data-testid="avatar"
/>
`;

exports[`Avatar should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<span
  class="rmd-avatar custom-class-name"
  data-testid="avatar"
  style="color: red;"
/>
`;

exports[`Avatar should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<span
  class="rmd-avatar rmd-avatar--primary"
  data-testid="avatar"
  style=""
/>
`;

exports[`Avatar should render an img if a src or imgProps are provided 1`] = `
<span
  class="rmd-avatar"
  data-testid="avatar"
>
  <img
    alt=""
    class="rmd-avatar__image"
    src="/some-image.png"
  />
</span>
`;

exports[`Avatar should render an img if a src or imgProps are provided 2`] = `
<span
  class="rmd-avatar"
  data-testid="avatar"
>
  <img
    alt=""
    class="rmd-avatar__image"
    referrerpolicy="no-referrer"
    src="/some-image.png"
  />
</span>
`;

exports[`Avatar should render an img if a src or imgProps are provided 3`] = `
<span
  class="rmd-avatar"
  data-testid="avatar"
>
  <img
    alt=""
    class="rmd-avatar__image custom-class-name"
    src="/another-image.png"
    style="opacity: 0.5;"
  />
</span>
`;

exports[`Avatar styling utils should allow the avatarImage to be called without any parameters 1`] = `"rmd-avatar__image"`;
